onnxruntime/python/session.py

```diff
@@ -303,3 +330,80 @@ def clear_binding_inputs(self):
 
     def clear_binding_outputs(self):
         self._iobinding.clear_binding_outputs()
+
+
+class OrtValue:
+    '''
+    This class provides APIs to construct and deal with OrtValues.
+    '''
+    def __init__(self, ortvalue):
+        if isinstance(ortvalue, C.OrtValue):
+            self._ortvalue = ortvalue
+        else:
+            # An end user won't hit this error
+            raise ValueError("`Provided ortvalue` needs to be of type " +
+                             "`onnxruntime.capi.onnxruntime_pybind11_state.OrtValue`")
+
+    @staticmethod
+    def tensor_from_numpy(numpy_obj, device_type='cpu', device_id=0):
+        '''
+        Factory method to construct an OrtValue (which holds a Tensor) from a given Numpy object
+        A copy of the data in the Numpy object is held by the OrtValue
```

Comment:
>A copy of the data [](start = 8, length = 18)

Can we do without a copy?

Reply:
we gain nothing by doing so, actually. In most cases, the user will request that the OrtValue's memory reside on CUDA for which we need to ultimately allocate (and manage the life cycle of) memory. We may as well make a copy and de-couple from the life cycle of the numpy object for the CPU case. For CUDA - as already reasoned above, we have to maintain our own copy of the device memory.

Reply:
To de-couple the Tensor’s life cycle from the numpy array’s life cycle. Otherwise, we have to ensure that the numpy array is alive till the OrtValue is alive. This is not impossible, we can hang onto a reference of the numpy array in the OrtValue python class and that should ensure that the GC doesn’t destroy the numpy array as long as the OrtValue lives. But I wanted to simplify the buffer ownership and felt that the copy is not significant overhead given that it is “offline” and shouldn’t affect perf benchmarking too much and then we wouldn’t have to worry about the life cycle of the numpy array. But if you all feel strongly about it, we can directly use the numpy buffer if the device is cpu and try keeping the numpy array alive by the idea described above

Reply:
I suggest we don't do the copy. If you want to do a copy, create a new method (say) new_ortvalue() and document it explicitly in both cases which one does the copy and let the user choose. For reference take a look at https://pytorch.org/docs/stable/generated/torch.as_tensor.html#torch.as_tensor where it explicitly says 'if the data is an ndarray of the corresponding dtype and the device is the cpu, no copy will be performed'. Also, https://pytorch.org/docs/stable/generated/torch.from_numpy.html#torch.from_numpy.

Reply:
Ok - no copy for the case when the device is cpu

---
In reply to: [493285658](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r493285658) [](ancestors = 493285658)